### ISSUE TITLE
Impossible to use CMake's find_package to find C++ library if C library is installed as well

### ISSUE THREAD

**Issue by traversaro** [msgpack/msgpack-c#1043]
After the split discussed in https://github.com/msgpack/msgpack-c/issues/872, it seems that if both the C and the C++ library are installed, only the C library is found. This happens as both library install a configuration file called `msgpack-config.cmake`, even if there is not direct conflict as the files are installed in different directories.

## Detailed example

To give you an example, I have a sample program like the following:
~~~
project(HelloWorld)
cmake_minimum_required(VERSION 3.16)

find_package(msgpack REQUIRED)

message(STATUS "msgpack_DIR: ${msgpack_DIR}")

if(TARGET msgpackc)
  message(STATUS "TARGET msgpackc is found")
else()
  message(STATUS "TARGET msgpackc is NOT found")
endif()

if(TARGET msgpackc-cxx)
  message(STATUS "TARGET msgpackc-cxx is found")
else()
  message(STATUS "TARGET msgpackc-cxx is NOT found")
endif()
~~~

If I run CMake when only the C++ library is installed in the `/home/traversaro/msgpack_ws/install` prefix, I obtained the following output:
~~~
traversaro@IITICUBLAP257:~/msgpack_ws/example/build$ cmake .
-- msgpack_DIR: /home/traversaro/msgpack_ws/install/lib/cmake/msgpackc-cxx
-- TARGET msgpackc is NOT found
-- TARGET msgpackc-cxx is found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/traversaro/msgpack_ws/example/build
~~~

However, if I also install the C library in  the `/home/traversaro/msgpack_ws/install` prefix, the following happens, i.e. the C library is found instead, making it impossible to use the `msgpackc-cxx` target:
~~~
-- msgpack_DIR: /home/traversaro/msgpack_ws/install/lib/cmake/msgpack
-- TARGET msgpackc is found
-- TARGET msgpackc-cxx is NOT found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/traversaro/msgpack_ws/example/build
~~~

**Comment by traversaro** [comment]
The most straightforward way to fix this is to modify either the C or C++ library to change their "CMake package name", i.e. the name that you pass to `find_package` to find it. However, this would mean breaking compatibility with existing downstream libraries that use `find_package(msgpack REQUIRED)`  to find the library that changed its "CMake package name".

**Comment by traversaro** [comment]
See also https://github.com/msgpack/msgpack-c/issues/974 that was kind of related to this.

**Comment by redboltz** [comment]
I'm not an expert of cmake.
Where the package name decided? Just a file name of `msgpack-config.cmake.in` ?
Now, both c_master and cpp_master has the same file `msgpack-config.cmake.in`. Is it wrong?

I didn't know user writes find_package as follows:

```cmake
find_package(msgpack REQUIRED)
```

It should be 

```cmake
find_package(msgpackc REQUIRED)
```

for C and

```cmake
find_package(msgpackc-cxx REQUIRED)
```

for C++.

```cmake
ADD_LIBRARY (msgpackc-cxx INTERFACE)
```

It is corresponding to 

```cmake
ADD_LIBRARY (msgpackc-cxx INTERFACE)
```

But for C 

```cmake
    ADD_LIBRARY (msgpackc
        ${msgpackc_SOURCES}
        ${msgpackc_HEADERS}
    )
```

and

```cmake
    ADD_LIBRARY (msgpackc-static STATIC
        ${msgpackc_SOURCES}
        ${msgpackc_HEADERS}
    )
```

are defined conditionally.

I think that `msgpackc` is acceptable name for C package.

After this update, user can write

```cpp
find_package(msgpackc REQUIRED)  # for C
find_package(msgpackc-cxx REQUIRED) # for C++
```

I'm not sure how user can get cmake variables such as C include directory and C++ include directory. I guess that cmake automatically defines the variables that has `msgpackc` `msgpackc-cxx` prefix. Is that right?

**Comment by redboltz** [comment]
I mean the current HEAD doesn't support

```cmake
find_package(msgpackc REQUIRED)  # for C
find_package(msgpackc-cxx REQUIRED) # for C++
```

Some modification is required. Reanme `msgpack-config.cmake.in` file and using code? I guess.
Then it should work. As you mentioned, it is a breaking change but user needs to accept the change.

Could you write PR for that ?

**Comment by traversaro** [comment]
> I'm not an expert of cmake.
> Where the package name decided? Just a file name of `msgpack-config.cmake.in` ?
> Now, both c_master and cpp_master has the same file `msgpack-config.cmake.in`.
> Am I understanding correctly?

Yes, that is correct, the `find_package(msgpack REQUIRED)` looks for a file named `msgpack-config.cmake` or `msgpackConfig.cmake` in many locations, and returns the first that it finds. If only one is found in the install prefix, there is no problem, but if both are found, only the first one found is returned.

**Comment by traversaro** [comment]
> I guess that cmake automatically defines the variables that has `msgpackc` `msgpackc-cxx` prefix. Is that right?

Actually all this informations are contained in the imported targets `msgpackc` and `msgpackc-cxx`, i.e. to compile a simple program you do:

~~~
project(msgpackcexample)

cmake_minimum_required(VERSION 3.16)

find_package(msgpack REQUIRED)

add_executable(msgpackcexample msgpackcexample.c)
target_link_libraries(msgpackcexample PRIVATE msgpackc)
~~~

or for C++:

~~~
project(msgpacxxkcexample)

cmake_minimum_required(VERSION 3.16)

find_package(msgpack REQUIRED)

add_executable(msgpackcxxexample msgpackcxxexample.c)
target_link_libraries(msgpackcxxexample PRIVATE msgpackc-cxx)
~~~

CMake automatically handles that the include directories, the definition and the libraries to link are correctly propagated.

> Could you write PR for that ?

Sure!

**Comment by traversaro** [comment]
I think that we can close as this was fixed in releases 5 of both the C and C++ library, thanks a lot!

**Comment by eli-schwartz** [comment]
I'm pretty confused why it was renamed to msgpackc-cxx, instead of, say, msgpack-cxx. It confused me twice when trying to figure out the new dependency name.

I guess the bigger issue, though, is that the C library is now called *both* msgpack (pkg-config), and msgpackc (cmake). I wonder why that was necessary at all. It should definitely be forced to be consistent, though.

**Comment by redboltz** [comment]
@traversaro could you answer the question?

I guess that C-library should be called msgpackc on pkg-config. But I don't have much knowledge about pkg-config.
https://github.com/msgpack/msgpack-c/tree/c_master has pkg-config related code?

**Comment by eli-schwartz** [comment]
Indeed:

https://github.com/msgpack/msgpack-c/blob/551308ae1068be1d122ec20e319fc9ead7900747/CMakeLists.txt#L117-L121
https://github.com/msgpack/msgpack-c/blob/551308ae1068be1d122ec20e319fc9ead7900747/CMakeLists.txt#L270-L272

This can be found with Meson using:
```meson
dependency('msgpack')
```

And the cmake files can also be found with meson:
```meson
dependency('msgpackc')
```

If they had the same name, then you could just look that up, and Meson would try both pkg-config and cmake.

...

By the way, maybe a pkg-config file for msgpackc-cxx could be provided?

**Comment by eli-schwartz** [comment]
For C:
- msgpack-c: name of repo, name of release tarball
- msgpack: name of pkg-config
- msgpackc: name of cmake project, name of cmake dependency files


For C++:
- msgpack-cxx: name of release tarball
- msgpackc-cxx: name of cmake project, name of cmake dependency files

I think there should be a unified naming scheme.

**Comment by redboltz** [comment]
Thans you!
I agree.

**Comment by redboltz** [comment]
I will write the PR for C and C++. It would be 6.0.0 (because of including breaking changes).
Then I request review to you.

**Comment by redboltz** [comment]
@eli-schwartz , could you review #1053 #1054 ?

These PR unify as follows:

For C: msgpack-c
For C++: msgpack-cxx

**Comment by traversaro** [comment]
> @traversaro could you answer the question?

In my PRs I followed the original suggestions for the CMake package names done in https://github.com/msgpack/msgpack-c/issues/1043#issuecomment-1372151371 . I agree that the naming suggested by @eli-schwartz is more clear.

**Comment by redboltz** [comment]
@traversaro Thank you for the comment. Yes I had decided to use `msgpackc-cxx` at that time.
Now, I'm going to update as https://github.com/msgpack/msgpack-c/issues/1043#issuecomment-1447535967

I worried about `msgpack-cxx` would have been used by other project, but as far as I've checked, it is not used.
( Some of existing package name is `msgpack-cxx` but it is for https://github.com/msgpack/msgpack-c/. so no problem)
